miner/block.go

```diff
@@ -0,0 +1,397 @@
+	// wait for the timestamp of header, use this to adjust the block period
+	delay := time.Until(time.Unix(int64(b.header.Time), 0))
+	select {
+	case <-time.After(delay):
+	case <-ctx.Done():
+		return
+	}
+
+	err = b.selectAndApplyTransactions(ctx, w)
+	if err != nil {
+		log.Error("Failed to apply transactions to the block", "err", err)
+		return
+	}
+	w.updateSnapshot(b)
+
+	block, err := b.finalizeAndAssemble(w)
+	if err != nil {
+		log.Error("Failed to finalize and assemble the block", "err", err)
+		return
+	}
+	w.updateSnapshot(b)
+	if w.isRunning() {
+		if w.fullTaskHook != nil {
+			w.fullTaskHook()
+		}
+		w.submitTaskToEngine(&task{receipts: b.receipts, state: b.state, block: block, createdAt: time.Now()})
+
+		feesEth := totalFees(block, b.receipts)
+		log.Info("Commit new mining work", "number", block.Number(), "sealhash", w.engine.SealHash(block.Header()),
+			"txs", b.tcount, "gas", block.GasUsed(), "fees", feesEth, "elapsed", common.PrettyDuration(time.Since(start)))
+
+	}
+
+}
+
+// prepareBlock intializes a new blockState that is ready to have transaction included to.
+func (w *worker) prepareBlock() (*blockState, error) {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	timestamp := time.Now().Unix()
+	parent := w.chain.CurrentBlock()
+
+	if parent.Time() >= uint64(timestamp) {
+		timestamp = int64(parent.Time() + 1)
+	}
+
+	num := parent.Number()
+	header := &types.Header{
+		ParentHash: parent.Hash(),
+		Number:     num.Add(num, common.Big1),
+		Extra:      w.extra,
+		Time:       uint64(timestamp),
+	}
+
+	txFeeRecipient := w.txFeeRecipient
+	if !w.chainConfig.IsDonut(header.Number) && w.txFeeRecipient != w.validator {
+		txFeeRecipient = w.validator
+		log.Warn("TxFeeRecipient and Validator flags set before split etherbase fork is active. Defaulting to the given validator address for the coinbase.")
+	}
+
+	// Only set the coinbase if our consensus engine is running (avoid spurious block rewards)
+	if w.isRunning() {
+		if txFeeRecipient == (common.Address{}) {
+			return nil, errors.New("Refusing to mine without etherbase")
+		}
+		header.Coinbase = txFeeRecipient
+	}
+	// Note: The parent seal will not be set when not validating
+	if err := w.engine.Prepare(w.chain, header); err != nil {
+		log.Error("Failed to prepare header for mining", "err", err)
+		return nil, fmt.Errorf("Failed to prepare header for mining: %w", err)
+	}
+
+	// Initialize the block state itself
+	state, err := w.chain.StateAt(parent.Root())
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get the parent state: %w:", err)
+	}
+
+	b := &blockState{
+		signer:         types.NewEIP155Signer(w.chainConfig.ChainID),
+		state:          state,
+		tcount:         0,
+		gasLimit:       core.CalcGasLimit(parent, state),
+		header:         header,
+		txFeeRecipient: txFeeRecipient,
+	}
+	b.gasPool = new(core.GasPool).AddGas(b.gasLimit)
+
+	// Play our part in generating the random beacon.
+	if w.isRunning() && random.IsRunning() {
```

Comment:
not part of this PR, but i would love to see all randomenss related code in a single module, which is being called from the different places we call it today (here & blockchain)